src/main/java/com/google/devtools/build/lib/actions/Artifact.java

```diff
@@ -648,16 +658,37 @@ public final Label getOwner() {
    * package-path entries (for source Artifacts), or one of the bin, genfiles or includes dirs (for
    * derived Artifacts). It will always be an ancestor of getPath().
    */
-  @Override
   public final ArtifactRoot getRoot() {
     return root;
   }
 
+  @Override
+  public final FileRootApi getRootForStarlark(StarlarkSemantics semantics) {
+    // It would *not* be correct to just apply PathMapper#map to the exec path of the root: The
+    // root part of the mapped exec path of this artifact may depend on its complete exec path as
+    // well as on e.g. the digest of the artifact.
+    PathFragment mappedExecPath = PathMapper.loadFrom(semantics).map(execPath);
```

Comment:
My professional opinion is that I don't know. What I *think* is that this shouldn't matter much, because `.root.` is not very frequently used and thus an extra map lookup shouldn't matter much. When in doubt, I'd rather err on the side of simple code.